src/einsteinpy/utils/dual.py

```diff
@@ -0,0 +1,222 @@
+
+    def sin(self):
+        return DualNumber(np.sin(self.a), self.b * np.cos(self.a))
+
+    def cos(self):
+        return DualNumber(np.cos(self.a), -self.b * np.sin(self.a))
+
+    def tan(self):
+        return np.sin(self) / np.cos(self)
+
+    def log(self):
+        return DualNumber(np.log(self.a), self.b / self.a)
+
+    def exp(self):
+        return DualNumber(np.exp(self.a), self.b * np.exp(self.a))
+
+
+def _deriv(func, x):
+    """
+    Calculates first (partial) derivative of ``func`` at ``x``
+
+    Parameters
+    ----------
+    func : callable
+        Function to differentiate
+    x : float
+        Point, at which, the derivative will be evaluated
+
+    Returns
+    _______
+    float
+        First partial derivative of ``func`` at ``x``
+
+    """
+    funcdual = func(DualNumber(x, 1.0))
+
+    if isinstance(funcdual, DualNumber):
+        return funcdual.b
+
+    return 0.0
+
+
+def _diff_g(g, g_prms, coords, indices, wrt):
+    """
+    Computes derivative of metric elements
+
+    Parameters
+    ----------
+    g : callable
+        Metric (Contravariant) Function
+    coords : array_like
+        4-Position
+    indices : array_like
+        2-tuple, containing indices, indexing a metric
+        element, whose derivative will be calculated
+    wrt : int
+        coordsinate, with respect to which, the derivative
+        will be calculated
+        Takes values from ``[0, 1, 2, 3]``
+
+    Returns
+    -------
+    float
+        Value of derivative of metric element at ``coords``
+
+    Raises
+    ------
+    ValueError
+        If ``wrt`` is not in [1, 2, 3, 4]
+        or ``len(indices) != 2``
+
+    """
+    if wrt not in [0, 1, 2, 3]:
+        raise ValueError(f"wrt takes values from [0, 1, 2, 3]. Supplied value: {wrt}")
+
+    if len(indices) != 2:
+        raise ValueError("indices must be a 2-tuple containing indices for the metric.")
+
+    dual_coords = [
+        DualNumber(coords[0], 0.0),
+        DualNumber(coords[1], 0.0),
+        DualNumber(coords[2], 0.0),
+        DualNumber(coords[3], 0.0),
+    ]
+
+    dual_coords[wrt].b = 1.0
+
+    return _deriv(lambda q: g(dual_coords, *g_prms)[indices], coords[wrt])
+
+
+def _jacobian_g(g, g_prms, coords, wrt):
+    """
+    Part of Jacobian of Metric
+
+    Parameters
+    ----------
+    g : callable
+        Metric (Contravariant) Function
+    g_prms : array_like
+        Tuple of parameters to pass to the metric
+        E.g., ``(a,)`` for Kerr
+    coords : array_like
+        4-Position
+    wrt : int
+        coordsinate, with respect to which, the derivative
+        will be calculated
+        Takes values from ``[0, 1, 2, 3]``
+
+    Returns
+    -------
+    numpy.ndarray
+        Value of derivative of metric elements,
+        w.r.t a particular coordinate, at ``coords``
+
+    """
+    J = np.zeros((4, 4))
+
+    for i in range(4):
+        for j in range(4):
```

Comment:
Isn't this calling `g` redundantly? The complete output metric will be computed in one pass in `_deriv`, then index is used to fetch one value from the metric. But then next time, we will get the same output metric computed and will index some other position?

Reply:
`_deriv` computes the derivative of a particular metric component, as marked by `[indices]` in this line (in `_diff_g`):
```
return _deriv(lambda q: g(dual_coords, *g_prms)[indices], coords[wrt])
```
As you would have noticed, `_deriv` does not loop over the metric. So, it can only differentiate 1 elementary function at a time. It's basically the first order partial derivative operator.

So, I have to loop over the rows and the columns, while keeping `wrt` (coordinate) fixed to get part of the Metric Jacobian against a specific coordinate.

I plan to profile loops vs numpy assignments to see, which one's faster. Will add the better solution to the next commit.

Reply:
* Current code with 2 `for` loops (All elements):
```
648 µs ± 9.39 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
* 1 `for` loop (Looping along rows):
```
1.01 ms ± 20.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
* 0 `for` loop:
```
650 µs ± 7.72 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
I suppose, we can keep the current code. It's nearly as efficient as with 0 loops.